policykit/policyengine/engine.py

```diff
@@ -0,0 +1,309 @@
+import logging
+
+from actstream import action as actstream_action
+from django.conf import settings
+
+from policyengine.utils import ActionKind
+
+logger = logging.getLogger(__name__)
+db_logger = logging.getLogger("db")
+
+
+class EvaluationLogAdapter(logging.LoggerAdapter):
+    def process(self, msg, kwargs):
+        kwargs["extra"] = self.extra
+        return (msg, kwargs)
+
+
+class EvaluationContext:
+    """
+    Class to hold all variables available in a policy evaluation.
+    All attributes on this class are in scope and can be used by the policy author.
+
+    Attributes:
+        proposal (Proposal): The proposal representing this evaluation.
+        action (BaseAction): The action that triggered this policy evaluation.
+        policy (Policy): The policy being evaluated.
+        slack (SlackCommunity)
+        discord (DiscordCommunity)
+        discourse (DiscourseCommunity)
+        reddit (RedditCommunity)
+        metagov (Metagov): Metagov library for performing enabled actions and processes.
+        logger (logging.Logger): Logger that will log messages to the PolicyKit web interface.
+
+    """
+
+    def __init__(self, proposal):
+        self.action = proposal.action
+        self.policy = proposal.policy
+        self.proposal = proposal
+        self.logger = EvaluationLogAdapter(
+            db_logger, {"community": proposal.action.community.community, "proposal": proposal}
+        )
+
+        from policyengine.models import Community, CommunityPlatform
+
+        parent_community: Community = self.action.community.community
+
+        # Make all CommunityPlatforms available in the evaluation context
+        for comm in CommunityPlatform.objects.filter(community=parent_community):
+            setattr(self, comm.platform, comm)
```

Comment:
This is the key change that makes policies now be "cross platform." For example in a Slack policy I could say `discord.post_message(...)` or `discord.initiate_vote(...)`, if I had a discord community connected to my slack community.